### ISSUE TITLE
App crash when implementing custom ios App Delegate

### ISSUE THREAD

**Issue by Macarthurval** [NativeScript/NativeScript#9045]
**Environment**
Provide version numbers for the following components (information can be retrieved by running `tns info` in your project folder or by inspecting the `package.json` of the project):
 - CLI: 7.0.11
 - Cross-platform modules: 7.0.13
 - iOS Runtime: 7.0.5
 - XCode Version: 12.2
 - Plugin(s): none

**Describe the bug**
When implementing a custom app delegate in iOS, the application crashes on startup. Here the error log:

	====== Assertion failed ======
	Native stack trace:
	1          0x101b692af tns::Assert(bool, v8::Isolate*) + 119
	2          0x101b8f5a8 tns::Interop::WriteValue(v8::Local<v8::Context>, tns::TypeEncoding const*, void*, v8::Local<v8::Value>) + 3156
	3          0x101b8e93c tns::Interop::SetFFIParams(v8::Local<v8::Context>, tns::TypeEncoding const*, tns::FFICall*, int, int, tns::V8Args&) + 110
	4          0x101b8e5a1 tns::Interop::CallFunctionInternal(tns::MethodCall&) + 421
	5          0x101b2607a tns::MetadataBuilder::CFunctionCallback(v8::FunctionCallbackInfo<v8::Value> const&) + 240
	6          0x101cb613c v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo) + 620
	7          0x101cb55ec v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) + 556
	8          0x101cb4c73 v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) + 259
	9          0x10254dc19 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit + 57
	JavaScript stack trace:
	at run (file: node_modules/@nativescript/core/application/index.ios.js:331:0)
	at <anonymous> (file:///app/bundle.js:160:64)
	at ./app.ts (file:///app/bundle.js:176:30)
	at __webpack_require__ (file: app/webpack/bootstrap:816:0)
	at checkDeferredModules (file: app/webpack/bootstrap:43:0)
	at webpackJsonpCallback (file: app/webpack/bootstrap:30:0)
	at <anonymous> (file:///app/bundle.js:2:57)
	at require (VM)

**To Reproduce**
- `ns create test`
- select _Plain TypeScript_ and _Hello World_
- `cd test`
- Add custom empty iOS app delegate in app.ts:

        import { ios } from '@nativescript/core/application';

        class MyDelegate extends UIResponder implements UIApplicationDelegate {
            public static ObjCProtocols = [UIApplicationDelegate];

            applicationDidFinishLaunchingWithOptions(
                application: UIApplication,
                launchOptions: NSDictionary<any, any>
            ): boolean {
                console.log( "applicationWillFinishLaunchingWithOptions: " + launchOptions );

                return true;
            }

            applicationDidBecomeActive(application: UIApplication): void {
                console.log("applicationDidBecomeActive: " + application);
            }
        }
        ios.delegate = MyDelegate;

- `ns run ios`

**Expected behavior**

Application launch without errors.

**Sample project**

**Additional context**

**Comment by shirakaba** [comment]
Have you tried adding the `@NativeClass` decorator to the class declaration?

**Comment by Macarthurval** [comment]
Omg, love u!, that does the trick. I've been stuck with this for two days 😅

Maybe will be a good idea to update the example for typescript app delegate at the documentation here:

[https://docs.nativescript.org/core-concepts/application-lifecycle#ios-uiapplicationdelegate](https://docs.nativescript.org/core-concepts/application-lifecycle#ios-uiapplicationdelegate)

Thanks!

**Comment by jonathan-casarrubias** [comment]
Hey guys, I was having same issue trying to create a Delegate that was making the app to crash until I updated to NS7 just to use NativeClass decorator, but I'm having the issue that decorated classes with @NativeClass works if I place those within my index.ios.ts file, but if I try to import from an external file, it would fail:

This works:

```js
import { GridLayout } from '@nativescript/core';

export class Test extends GridLayout {
    constructor() {
        super();
        this.init();
    }

    init() {
        if (this.nativeView) {
            console.log('Initializing Test');
            const test: TestViewController = <TestViewController> TestViewController.alloc().init();
            test.view.frame = this.nativeView.bounds;
            test.view.backgroundColor = UIColor.blackColor;
            this.nativeView.insertSubviewAtIndex(test.view, 0);
        } else {
            setTimeout(() => this.init(), 500);
        }
    }
}


@NativeClass()
export class TestViewController extends UIViewController {
    viewDidLoad() {
        console.log('A NATIVE VIEW LOADED');
    }
}
```

While this does not

```js
import { GridLayout } from '@nativescript/core';
import { TestViewController } from './ios/views/test.view'; // once imported starts to fail

export class Test extends GridLayout {
    constructor() {
        super();
        this.init();
    }

    init() {
        if (this.nativeView) {
            console.log('Initializing Test');
            const test: TestViewController = <TestViewController> TestViewController.alloc().init();
            test.view.frame = this.nativeView.bounds;
            test.view.backgroundColor = UIColor.blackColor;
            this.nativeView.insertSubviewAtIndex(test.view, 0);
        } else {
            setTimeout(() => this.init(), 100);
        }
    }
}
```

![image](https://user-images.githubusercontent.com/1533239/99877456-0147d000-2bc4-11eb-8462-e8b3ca7ca6e8.png)

Any ideas on this one?

**Comment by shirakaba** [comment]
@jonathan-casarrubias just a guess, but instead of writing:

```
@NativeClass 
export class TestViewController // ...
```

... Write:


```
@NativeClass 
class TestViewController // ...

export { TestViewController };
```

I saw a GitHub gist about this odd limitation. @farfromrefug or @NathanWalker may be able to explain.

**Comment by jonathan-casarrubias** [comment]
@shirakaba ok very interesting, I've just tried and it did work!!! I believe this:

https://nativescript.org/blog/nativescript-7-for-plugin-authors/

Might be misleading, but anyways is working now.. Thanks a lot  🙌

**Comment by shirakaba** [comment]
Good point about those docs. @NathanWalker could that blog post please be edited to reflect how `export` should not be written inline when using `@NativeClass`?

Otherwise, I do believe this issue is closed!

**Comment by farfromrefug** [comment]
@shirakaba i see the issue . the nativeclass transformer needs a fix to handle that case

**Comment by jonathan-casarrubias** [comment]
@farfromrefug @shirakaba thanks guys, I have one more question I'm not sure I should open a new issue or note, so will ask the question here and will follow your command, if I create a new story of it is something easy..

So after migrating and doing the above ^^^ I stopped getting the errors I had, but it seems that the methods declared within the delegate class are not being executed.

Oddly if I remove the methods I get an error only for one of the methods (though many are required) so I get the following

```sh
-[RTCPeerConnectionDelegateImpl peerConnectionShouldNegotiate:]: unrecognized selector sent to instance 0x2831e8e70
```

If I return the class to where the methods are declared, the error disappear but is like nothing happens, I Get no logs or anything executed within the delegated methods,

thoughts?

**Comment by shirakaba** [comment]
@jonathan-casarrubias For any methods that are not already exposed by the protocol you’ve specified, you need to explicitly expose them. See `ObjCExposedMethods` in:

https://docs.nativescript.org/core-concepts/ios-runtime/how-to/objc-subclassing#typescript-support

See also the “limitations” section. You cannot expose static methods. So if initWithClientDelegate is not already exposed by that protocol, and you really need it to be exposed, then you’re out of luck and need to write a native class instead.

**Comment by jonathan-casarrubias** [comment]
@shirakaba InitWithClientDelegate is not required to be exposed, only those with the peerConnection naming, even if I remove that and I only use something like this (which is the actual protocol) nothing happens, I assume I should see those logs in the terminal

```js
@NativeClass()
class RTCPeerConnectionDelegateImpl extends NSObject implements RTCPeerConnectionDelegate {

  public static ObjCProtocols = [RTCPeerConnectionDelegate];

  static new(): RTCPeerConnectionDelegateImpl {
    return <RTCPeerConnectionDelegateImpl>super.new();
  }

  peerConnectionDidAddReceiverStreams?(peerConnection: RTCPeerConnection, rtpReceiver: RTCRtpReceiver, mediaStreams: NSArray<RTCMediaStream> | RTCMediaStream[]): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidAddStream(peerConnection: RTCPeerConnection, stream: RTCMediaStream): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidChangeConnectionState?(peerConnection: RTCPeerConnection, newState: RTCPeerConnectionState): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidChangeIceConnectionState(peerConnection: RTCPeerConnection, newState: RTCIceConnectionState): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidChangeIceGatheringState(peerConnection: RTCPeerConnection, newState: RTCIceGatheringState): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidChangeSignalingState(peerConnection: RTCPeerConnection, stateChanged: RTCSignalingState): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidGenerateIceCandidate(peerConnection: RTCPeerConnection, candidate: RTCIceCandidate): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidOpenDataChannel(peerConnection: RTCPeerConnection, dataChannel: RTCDataChannel): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidRemoveIceCandidates(peerConnection: RTCPeerConnection, candidates: NSArray<RTCIceCandidate> | RTCIceCandidate[]): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidRemoveReceiver?(peerConnection: RTCPeerConnection, rtpReceiver: RTCRtpReceiver): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidRemoveStream(peerConnection: RTCPeerConnection, stream: RTCMediaStream): void {
    console.log("Method not implemented.");
  }
  peerConnectionDidStartReceivingOnTransceiver?(peerConnection: RTCPeerConnection, transceiver: RTCRtpTransceiver): void {
    console.log("Method not implemented.");
  }
  peerConnectionShouldNegotiate(peerConnection: RTCPeerConnection): void {
    console.log("peerConnection should negotiate");
  }
}

export { RTCPeerConnectionDelegateImpl };
```

**Comment by shirakaba** [comment]
@jonathan-casarrubias Instead of creating `RTCPeerConnectionDelegateImpl.new()` inline and passing it in directly as a construction param, try storing a reference to it (somewhere that won’t go out of scope – e.g. `this.myDelegate = RTCPeerConnectionDelegateImpl.new()`) and passing that reference in. This should keep it retained (have to be careful about strong/weak native refs even in NativeScript).

I’m not sure you need to write `@NativeClass()`, either; `@NativeClass` should suffice.

**Comment by shirakaba** [comment]
@jonathan-casarrubias Could you also try `RTCPeerConnectionDelegateImpl.alloc().init()` instead of `RTCPeerConnectionDelegateImpl.new()`?

**Comment by jonathan-casarrubias** [comment]
@shirakaba I will try those right now, I'll try one by one and see if there is a difference which one made it... Will update with the results

**Comment by jonathan-casarrubias** [comment]
@shirakaba ok so here are my results, hopefully helpful for anyone in the future digging in this issue..

So as a context I started by the one that made the more sense because of the diagnosis, it did sound like a lost reference so I started with that one and the sdk started working again, I still don't see any log but I believe I need to negotiate a connection first in order to receive those, but once setting a local reference `this.peerConnectionDelegate = ....` everything seems to correctly work again, for the rest I need to check on my side first, so my results are:

1.- Once setting a local reference the SDK works as expected, at least it won't hang and the rest of the logic after works, so I still don't get logs but I think is on my side.
2.- Changing from @NativeScript() to @NativeScript and backwards does not make any difference.
3.- Changing from .new to .alloc().init() and backwards does not make any difference.

Both test 2 and 3, were tested after the sdk started working again without presenting a difference in terms of making the sdk to fail or not, it will just keep working as long as number 1 is covered.

I'll take it from here thanks a lot for your help @shirakaba  🙌
